src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs

```diff
@@ -8,5 +13,70 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
         {
             return (BoundExpression)Visit(node);
         }
+
+        public override BoundNode VisitBinaryOperator(BoundBinaryOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.ConstantValueOpt, node.MethodOpt, node.ResultKind, left, right, type));
+        }
+
+        public override BoundNode VisitUserDefinedConditionalLogicalOperator(BoundUserDefinedConditionalLogicalOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.LogicalOperator, node.TrueOperator, node.FalseOperator, node.ResultKind, left, right, type));
+        }
+
+        private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator, Func<BoundBinaryOperatorBase, BoundExpression, BoundExpression, TypeSymbol, BoundBinaryOperatorBase> nodeUpdater)
+        {
+            // Use an explicit stack to avoid blowing the managed stack when visiting deeply-recursive
+            // binary nodes
+            var stack = ArrayBuilder<BoundBinaryOperatorBase>.GetInstance();
+            BoundBinaryOperatorBase currentBinary;
+            BoundExpression child = binaryOperator;
+
+            while (true)
+            {
+                if (child is BoundBinaryOperatorBase childBinary)
+                {
+                    currentBinary = childBinary;
+                    child = childBinary.Left;
+                    stack.Push(currentBinary);
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            Debug.Assert(stack.Count > 0);
+            currentBinary = null;
+
+            while (true)
+            {
+                if (stack.Count == 0)
+                {
+                    return currentBinary;
+                }
+
+                var left = currentBinary;
+                currentBinary = stack.Pop();
+                currentBinary = updateNode(currentBinary, left, this, nodeUpdater);
+            }
+
+            static BoundBinaryOperatorBase updateNode(BoundBinaryOperatorBase node, BoundExpression leftOpt, NullabilityRewriter rewriter, Func<BoundBinaryOperatorBase, BoundExpression, BoundExpression, TypeSymbol, BoundBinaryOperatorBase> nodeUpdater)
```

Comment:
>updateNode [](start = 43, length = 10)

Perhaps make this an instance method, to avoid the `rewriter` argument.

Reply:
I'd prefer to keep this as a local function, as it's only relevant to this function.

---
In reply to: [269160348](https://github.com/dotnet/roslyn/pull/34429#discussion_r269160348) [](ancestors = 269160348)